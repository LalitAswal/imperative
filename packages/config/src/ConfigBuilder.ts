/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { IImperativeConfig } from "../../imperative";
import { Config } from "./Config";
import { IConfig } from "./doc/IConfig";
import { IConfigBuilderOpts } from "./doc/IConfigBuilderOpts";
import { IConfigProfile } from "./doc/IConfigProfile";

export class ConfigBuilder {
    private static readonly DEFAULT_ROOT_PROFILE_NAME = "my_profiles";

    /**
     * Build a new Config object from an Imperative CLI app configuration.
     * @param impConfig The Imperative CLI app configuration.
     * @param opts Options to control aspects of the builder.
     */
    public static async build(impConfig: IImperativeConfig, opts?: IConfigBuilderOpts): Promise<IConfig> {
        opts = opts || {};
        const config: IConfig = Config.empty();

        const baseProfileType: string = impConfig.baseProfile?.type;
        const rootProfileName: string = impConfig.templateProfileName || ConfigBuilder.DEFAULT_ROOT_PROFILE_NAME;

        for (const profile of impConfig.profiles) {
            let profileShortPath = `my_${profile.type}`;
            let profileLongPath = `profiles.${profileShortPath}`;
            if (baseProfileType && profile.type !== baseProfileType) {
                // Path should have two levels for non-base profiles
                profileShortPath = `${rootProfileName}.${profile.type}`;
                profileLongPath = `profiles.${rootProfileName}.profiles.${profile.type}`;
                lodash.set(config.profiles, `${rootProfileName}.properties`, {});
            }

            const properties: { [key: string]: any } = {};
            for (const [k, v] of Object.entries(profile.schema.properties)) {
                if (opts.populateProperties && v.includeInTemplate) {
                    if (v.secure) {
                        const propPath = `${profileLongPath}.properties.${k}`;
                        config.secure.push(propPath);

                        const propValue = opts.getSecureValue ? await opts.getSecureValue(k, v) : null;
                        if (propValue != null) {
                            // Save this value to be stored securely after profile is built
                            properties[k] = propValue;
                        }
                    } else {
                        if ((v as any).optionDefinition != null) {
                            // Use default value of ICommandOptionDefinition if present
                            properties[k] = (v as any).optionDefinition.defaultValue;
                        }
                        if (properties[k] === undefined) {
                            // Fall back to an empty value
                            properties[k] = this.getDefaultValue(v.type);
                        }
                    }
                }
            }

            // Add the profile to config and set it as default
            lodash.set(config, profileLongPath, {
                type: profile.type,
                properties
            });

            if (opts.populateProperties) {
                config.defaults[profile.type] = profileShortPath;
            }
        }

        // Hoist duplicate default properties
        if (config.profiles != null && config.profiles[rootProfileName] != null) {
            config.profiles[rootProfileName] = this.hoistTemplateProperties(config.profiles[rootProfileName]);
        }

        return config;
    }

    /**
     * Returns empty value that is appropriate for the property type.
     * @param propType The type of profile property
     * @returns Null or empty object
     */
    private static getDefaultValue(propType: string | string[]): any {
        // TODO How to handle profile property with multiple types
        if (Array.isArray(propType)) {
            propType = propType[0];
        }
        switch (propType) {
            case "string":  return "";
            case "number":  return 0;
            case "object":  return {};
            case "array":   return [];
            case "boolean": return false;
            default:        return null;
        }
    }

    /**
     * Moves duplicate default properties across child profiles up to root
     * profile.
     * @param rootProfile The root profile object of the config JSON
     * @returns The root profile with duplicate properties hoisted
     */
    private static hoistTemplateProperties(rootProfile: IConfigProfile): IConfigProfile {
        // Flatten properties into object that maps property name to list of values
        const flattenedProps: { [key: string]: any[] } = {};
        for (const childProfile of Object.values(rootProfile.profiles)) {
            for (const [k, v] of Object.entries(childProfile.properties)) {
                flattenedProps[k] = [...(flattenedProps[k] || []), v];
            }
        }
        // List property names defined multiple times with the same value
        const duplicateProps: string[] = [];
        for (const [k, v] of Object.entries(flattenedProps)) {
            if (v.length > 1 && (new Set(v)).size === 1) {
                duplicateProps.push(k);
            }
        }
        // Remove duplicate properties from child profiles and store them in root profile
        for (const propName of duplicateProps) {
            rootProfile.properties[propName] = flattenedProps[propName][0];
            for (const childProfile of Object.values(rootProfile.profiles)) {
                delete childProfile.properties[propName];
            }
        }
        return rootProfile;
    }
}
