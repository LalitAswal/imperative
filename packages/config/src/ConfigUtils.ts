/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ImperativeConfig } from "../../utilities";
import { ImperativeError } from "../../error";

export function secureSaveError(solution?: string): ImperativeError {
    const displayName = ImperativeConfig.instance.loadedConfig.productDisplayName || ImperativeConfig.instance.loadedConfig.name;
    let details = `Possible Solutions:\n` +
        ` 1. Reinstall ${displayName}. On Linux systems, also make sure to install the prerequisites listed in ${displayName} documentation.\n` +
        ` 2. Ensure ${displayName} can access secure credential storage. ${displayName} needs access to the OS to securely save credentials.`;
    if (solution != null) {
        details += `\n 3. ${solution}`;
    }
    return new ImperativeError({
        msg: "Unable to securely save credentials.",
        additionalDetails: details
    });
}
