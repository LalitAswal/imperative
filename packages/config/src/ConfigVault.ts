/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IConfigVault } from "./doc/IConfigVault";

export class ConfigVault {
    private _vault: IConfigVault;

    // Maximum credential length allowed on Windows 7 and newer
    private static readonly MAX_CREDENTIAL_LENGTH = 2560;

    /**
     * Initialize a config vault using the provided vault backend.
     * @param vault Backend with load and save methods
     */
    constructor(vault: IConfigVault) {
        this._vault = vault;
    }

    public async load(key: string): Promise<any> {
        // Load initial value from vault
        let value = await this._vault.load(key);

        // On Windows, try to load more values from vault
        if (process.platform === "win32" && value != null) {
            let index = 1;
            // Check if we've finished reading null-terminated JSON string
            while (!value.endsWith('\0')) {
                index++;
                // Load more values from vault and concat them
                value += await this._vault.load(`${key}-${index}`);
            }
            // Strip off trailing null char
            value = value.slice(0, -1);
        }

        return value;
    }

    public async save(key: string, value: string): Promise<void> {
        // Calculate max credential length before Base64 encoding
        const maxCredentialLength = Math.ceil(ConfigVault.MAX_CREDENTIAL_LENGTH * 0.75) - 4;
        if (process.platform === "win32" && value.length > maxCredentialLength) {
            value += '\0';
            let index = 1;
            while (value.length > 0) {
                const newKey = (index > 1) ? `${key}-${index}` : key;
                await this._vault.save(newKey, value.slice(0, maxCredentialLength));
                value = value.slice(maxCredentialLength);
                index++;
            }
        } else {
            await this._vault.save(key, value);
        }
    }
}
