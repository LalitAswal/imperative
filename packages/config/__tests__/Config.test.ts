/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { ImperativeError } from "../..";
import { Config } from "../src/Config";
import * as JSONC from "comment-json";

const MY_APP = "my_app";

describe("Config tests", () => {
    afterEach(() => {
        jest.restoreAllMocks();
    });

    describe("load", () => {
        beforeEach(() => {
            jest.spyOn(Config.prototype as any, "secureLoad").mockResolvedValue(undefined);
        });

        it("should load project user config", async () => {
            jest.spyOn(Config, "search").mockReturnValue(__dirname + "/__resources__/project.config.user.json");
            jest.spyOn(fs, "existsSync")
                .mockReturnValueOnce(true)      // Project user layer
                .mockReturnValueOnce(false)     // Project layer
                .mockReturnValueOnce(false)     // User layer
                .mockReturnValueOnce(false);    // Global layer
            const config = await Config.load(MY_APP);
            expect(config.properties).toMatchSnapshot();
        });

        it("should load project config", async () => {
            jest.spyOn(Config, "search").mockReturnValue(__dirname + "/__resources__/project.config.json");
            jest.spyOn(fs, "existsSync")
                .mockReturnValueOnce(false)     // Project user layer
                .mockReturnValueOnce(true)      // Project layer
                .mockReturnValueOnce(false)     // User layer
                .mockReturnValueOnce(false);    // Global layer
            const config = await Config.load(MY_APP);
            expect(config.properties).toMatchSnapshot();
        });

        it("should load user config", async () => {
            jest.spyOn(Config, "search").mockReturnValue(null);
            jest.spyOn(fs, "existsSync")
                .mockReturnValueOnce(false)     // Project user layer
                .mockReturnValueOnce(false)     // Project layer
                .mockReturnValueOnce(true)      // User layer
                .mockReturnValueOnce(false);    // Global layer
            const config = await Config.load(MY_APP, { homeDir: __dirname + "/__resources__" });
            expect(config.properties).toMatchSnapshot();
        });

        it("should load global config", async () => {
            jest.spyOn(Config, "search").mockReturnValue(null);
            jest.spyOn(fs, "existsSync")
                .mockReturnValueOnce(false)     // Project user layer
                .mockReturnValueOnce(false)     // Project layer
                .mockReturnValueOnce(false)     // User layer
                .mockReturnValueOnce(true);     // Global layer
            const config = await Config.load(MY_APP, { homeDir: __dirname + "/__resources__" });
            expect(config.properties).toMatchSnapshot();
        });

        it("should merge multiple config files", async () => {
            jest.spyOn(Config, "search")
                .mockReturnValueOnce(__dirname + "/__resources__/project.config.user.json")
                .mockReturnValueOnce(__dirname + "/__resources__/project.config.json");
            jest.spyOn(fs, "existsSync").mockReturnValue(true);
            const config = await Config.load(MY_APP, { homeDir: __dirname + "/__resources__" });
            expect(config.properties).toMatchSnapshot();
        });

        it("should load a config and populate missing defaults", async () => {
            jest.spyOn(Config, "search").mockReturnValue(__dirname + "/__resources__/project.config.json");
            jest.spyOn(fs, "existsSync")
                .mockReturnValueOnce(false)     // Project user layer
                .mockReturnValueOnce(true)      // Project layer
                .mockReturnValueOnce(false)     // User layer
                .mockReturnValueOnce(false);    // Global layer
            jest.spyOn(fs, "readFileSync").mockReturnValueOnce("{}");
            const config = await Config.load(MY_APP);
            expect(config.properties).toMatchSnapshot();
            expect(config.properties.defaults).toEqual({});
            expect(config.properties.profiles).toEqual({});
            expect(config.properties.plugins).toEqual([]);
            expect(config.properties.secure).toEqual([]);
        });

        it("should fail to load config that is not JSON", async () => {
            jest.spyOn(Config, "search").mockReturnValue(__dirname + "/__resources__/project.config.json");
            jest.spyOn(fs, "existsSync")
                .mockReturnValueOnce(false)     // Project user layer
                .mockReturnValueOnce(true)      // Project layer
                .mockReturnValueOnce(false)     // User layer
                .mockReturnValueOnce(false);    // Global layer
            jest.spyOn(fs, "readFileSync").mockReturnValueOnce("This is not JSON");
            let error: any;
            try {
                await Config.load(MY_APP);
            } catch (err) {
                error = err;
            }
            expect(error).toBeDefined();
            expect(error.message).toContain("error reading config file");
            expect(error.message).toContain(__dirname + "/__resources__");
            expect(error instanceof ImperativeError).toBe(true);
        });
    });

    it("should find config that exists if any layers exist", () => {
        const config = new (Config as any)();
        config._layers = [
            { exists: false },
            { exists: true },
            { exists: false }
        ];
        expect(config.exists).toBe(true);
    });

    it("should not find config that exists if no layers exist", () => {
        const config = new (Config as any)();
        config._layers = [ { exists: false } ];
        expect(config.exists).toBe(false);
    });

    it("should provide a deep copy of layers", () => {
        const config = new (Config as any)();
        config._layers = {};
        config.layers.properties = {};
        expect(Object.keys(config._layers).length).toBe(0);
    });

    describe("set", () => {
        beforeEach(() => {
            jest.spyOn(Config, "search").mockReturnValue(__dirname + "/__resources__/project.config.user.json");
            jest.spyOn(fs, "existsSync").mockReturnValueOnce(true).mockReturnValue(false);
            jest.spyOn(Config.prototype as any, "secureLoad").mockResolvedValue(undefined);
        });

        it("should set boolean true in config", async () => {
            const config = await Config.load(MY_APP);
            config.set("profiles.fruit.profiles.apple.properties.ripe", "true");
            expect(config.properties.profiles.fruit.profiles.apple.properties.ripe).toBe(true);
        });

        it("should set boolean false in config", async () => {
            const config = await Config.load(MY_APP);
            config.set("profiles.fruit.profiles.apple.properties.ripe", "false");
            expect(config.properties.profiles.fruit.profiles.apple.properties.ripe).toBe(false);
        });

        it("should set integer value in config", async () => {
            const config = await Config.load(MY_APP);
            config.set("profiles.fruit.profiles.apple.properties.price", "2");
            expect(config.properties.profiles.fruit.profiles.apple.properties.price).toBe(2);
        });

        it("should append to array value in config", async () => {
            const config = await Config.load(MY_APP);
            config.set("profiles.fruit.properties.tags", []);
            config.set("profiles.fruit.properties.tags", "sweet");
            expect(config.properties.profiles.fruit.properties.tags.length).toBe(1);
            expect(config.properties.profiles.fruit.properties.tags[0]).toBe("sweet");
        });

        it("should set secure string value in config", async () => {
            const config = await Config.load(MY_APP);
            const layer = (config as any).layerActive();
            config.set("profiles.fruit.profiles.apple.properties.secret", "@ppl3", { secure: true });
            expect(config.properties.profiles.fruit.profiles.apple.properties.secret).toBe("@ppl3");
            expect(layer.properties.secure.length).toBe(1);
            expect(layer.properties.secure[0]).toBe("profiles.fruit.profiles.apple.properties.secret");
        });

        it("should set schema URI at top of config", async () => {
            const config = await Config.load(MY_APP);
            const layer = (config as any).layerActive();
            config.setSchema("./schema.json");
            const jsonText = JSON.stringify(layer.properties);
            expect(jsonText.match(/^{\s*"\$schema":/)).not.toBeNull();
        });

        it("should save schema to disk if object is provided", async () => {
            const writeFileSpy = jest.spyOn(fs, "writeFileSync").mockReturnValue(undefined);
            const config = await Config.load(MY_APP);
            const schemaObj = { $schema: "./schema.json" };
            config.setSchema(schemaObj.$schema, schemaObj);
            expect(writeFileSpy).toHaveBeenCalledTimes(1);
            const jsonText = writeFileSpy.mock.calls[0][1];
            expect(jsonText).toBeDefined();
            expect(jsonText.match(/^{\s*"\$schema":/)).not.toBeNull();
        });

        it("should add new secure property to config", async () => {
            const config = await Config.load(MY_APP);
            const layer = (config as any).layerActive();
            const secureProp = "profiles.fruit.properties.secret";
            config.addSecure(secureProp);
            expect(layer.properties.secure.includes(secureProp)).toBe(true);
            config.addSecure(secureProp);
            expect(layer.properties.secure.filter((x: any) => x === secureProp).length).toBe(1);
        });

        it("should add a new layer when one is specified in the set", async () => {
            const config = await Config.load(MY_APP);
            config.set("profiles.fruit.profiles.mango.properties.color", "orange");
            expect(config.properties.profiles.fruit.profiles.mango.properties.color).toBe("orange");
            expect (config.properties.profiles).toMatchSnapshot();
        });
    });

    describe("set (with comments)", () => {
        const blockComment = "/* block-comment */";
        const lineComment = "// line-comment";
        beforeEach(() => {
            jest.spyOn(Config, "search").mockReturnValue(__dirname + "/__resources__/commented-project.config.user.json");
            jest.spyOn(fs, "existsSync").mockReturnValueOnce(true).mockReturnValue(false);
            jest.spyOn(Config.prototype as any, "secureLoad").mockResolvedValue(undefined);
        });

        it("should set boolean true in config", async () => {
            const config = await Config.load(MY_APP);
            config.set("profiles.fruit.profiles.apple.properties.ripe", "true");
            expect(config.properties.profiles.fruit.profiles.apple.properties.ripe).toBe(true);

            const layer = config.api.layers.get();
            expect(JSONC.stringify(layer.properties.profiles.fruit.profiles.apple.properties.ripe, null, Config.INDENT)).toContain(blockComment);
            expect(JSONC.stringify(layer.properties.profiles.fruit.profiles.apple.properties.ripe, null, Config.INDENT)).toContain(lineComment);
        });

        it("should set boolean false in config", async () => {
            const config = await Config.load(MY_APP);
            config.set("profiles.fruit.profiles.apple.properties.ripe", "false");
            expect(config.properties.profiles.fruit.profiles.apple.properties.ripe).toBe(false);

            const layer = config.api.layers.get();
            expect(JSONC.stringify(layer.properties.profiles.fruit.profiles.apple.properties.ripe, null, Config.INDENT)).toContain(blockComment);
            expect(JSONC.stringify(layer.properties.profiles.fruit.profiles.apple.properties.ripe, null, Config.INDENT)).toContain(lineComment);
        });

        it("should set integer value in config", async () => {
            const config = await Config.load(MY_APP);
            config.set("profiles.fruit.profiles.apple.properties.price", "2");
            expect(config.properties.profiles.fruit.profiles.apple.properties.price).toBe(2);

            const layer = config.api.layers.get();
            expect(JSONC.stringify(layer.properties.profiles.fruit.profiles.apple.properties.price, null, Config.INDENT)).toContain(blockComment);
            expect(JSONC.stringify(layer.properties.profiles.fruit.profiles.apple.properties.price, null, Config.INDENT)).toContain(lineComment);
        });

        it("should append to array value in config", async () => {
            const config = await Config.load(MY_APP);
            // config.set("profiles.fruit.properties.tags", []);
            config.set("profiles.fruit.properties.tags", "sweet");
            expect(config.properties.profiles.fruit.properties.tags.length).toBe(1);
            expect(config.properties.profiles.fruit.properties.tags[0]).toBe("sweet");

            const layer = config.api.layers.get();
            expect(JSONC.stringify(layer.properties.profiles.fruit.properties.tags, null, Config.INDENT)).toContain(blockComment);
            expect(JSONC.stringify(layer.properties.profiles.fruit.properties.tags, null, Config.INDENT)).toContain(lineComment);
        });

        it("should set secure string value in config", async () => {
            const config = await Config.load(MY_APP);
            const layer = (config as any).layerActive();
            layer.path = path.basename(layer.path); // Everyone has a different path.

            config.set("profiles.fruit.profiles.apple.properties.secret", "@ppl3", { secure: true });

            expect(config.properties.profiles.fruit.profiles.apple.properties.secret).toBe("@ppl3");
            expect(layer.properties.secure.length).toBe(1);
            expect(layer.properties.secure[0]).toBe("profiles.fruit.profiles.apple.properties.secret");

            expect(JSONC.stringify(layer.properties.secure, null, Config.INDENT)).toContain(blockComment);
            expect(JSONC.stringify(layer.properties.secure, null, Config.INDENT)).toContain(lineComment);
        });

        it("should set schema URI at top of config", async () => {
            const config = await Config.load(MY_APP);
            const layer = (config as any).layerActive();
            config.setSchema("./schema.json");
            const jsonText = JSONC.stringify(layer.properties);
            expect(jsonText.match(/^{\s*"\$schema":/)).not.toBeNull();

            expect(JSONC.stringify(layer.properties.$schema, null, Config.INDENT)).toContain(blockComment);
            expect(JSONC.stringify(layer.properties.$schema, null, Config.INDENT)).toContain(lineComment);
        });

        // it("should save schema to disk if object is provided", async () => {
        //     const writeFileSpy = jest.spyOn(fs, "writeFileSync").mockReturnValue(undefined);
        //     const config = await Config.load(MY_APP);
        //     const schemaObj = { $schema: "./schema.json" };
        //     config.setSchema(schemaObj.$schema, schemaObj);
        //     expect(writeFileSpy).toHaveBeenCalledTimes(1);
        //     const jsonText = writeFileSpy.mock.calls[0][1];
        //     expect(jsonText).toBeDefined();
        //     expect(jsonText.match(/^{\s*"\$schema":/)).not.toBeNull();
        // });

        // it("should add new secure property to config", async () => {
        //     const config = await Config.load(MY_APP);
        //     const layer = (config as any).layerActive();
        //     const secureProp = "profiles.fruit.properties.secret";
        //     config.addSecure(secureProp);
        //     expect(layer.properties.secure.includes(secureProp)).toBe(true);
        //     config.addSecure(secureProp);
        //     expect(layer.properties.secure.filter((x: any) => x === secureProp).length).toBe(1);
        // });

        // it("should add a new layer when one is specified in the set", async () => {
        //     const config = await Config.load(MY_APP);
        //     config.set("profiles.fruit.profiles.mango.properties.color", "orange");
        //     expect(config.properties.profiles.fruit.profiles.mango.properties.color).toBe("orange");
        //     expect (config.properties.profiles).toMatchSnapshot();
        // });
    });

    describe("paths", () => {
        beforeEach(() => {
            jest.spyOn(Config, "search").mockReturnValue(__dirname + "/__resources__/project.config.user.json");
            jest.spyOn(fs, "existsSync").mockReturnValueOnce(true).mockReturnValue(false);
            jest.spyOn(Config.prototype as any, "secureLoad").mockResolvedValue(undefined);
        });
        it("should get paths", async () => {
            const config = await Config.load(MY_APP);
            const paths: string[] = config.paths;
            const expectedPath: string = __dirname + "/__resources__/project.config.user.json";
            expect(paths).toContain(expectedPath);
        });
    });

    describe("search", () => {
        const home = os.homedir();
        const configFile = "project.config.user.json";
        const configDir = path.join(__dirname, "__resources__");
        const fakeConfigDir = path.join(__dirname, configFile);
        const parsedPath = path.parse(__dirname);
        it("should search for a file in the same directory", async () => {
            const joinedPath = path.join(configDir, configFile);
            const expectedPath = path.join(path.resolve(configDir), configFile);
            jest.spyOn(fs, "existsSync").mockReturnValue(true);
            jest.spyOn(path, "join").mockReturnValue(joinedPath);
            const file = Config.search(configFile, { stop: home });
            expect(file).toBe(expectedPath);
        });
        it("should search for a file in the parent directory", async () => {
            const joinedPath = path.join(configDir, configFile);
            const expectedPath = path.join(path.resolve(configDir, ".."), configFile);
            jest.spyOn(fs, "existsSync").mockReturnValueOnce(false).mockReturnValue(true);
            jest.spyOn(path, "join").mockReturnValue(joinedPath);
            const file = Config.search(configFile, { stop: home });
            expect(file).toBe(expectedPath);
        });
        it("should search for and not return file in the parent directory because the parent directory is the global directory", async () => {
            const joinedPath = path.join(configDir, configFile);
            const notExpectedPath = path.join(path.resolve(configDir, ".."), configFile);
            const expectedPath = path.join(path.resolve(configDir, "..", ".."), configFile);
            jest.spyOn(fs, "existsSync").mockReturnValueOnce(false).mockReturnValue(true);
            jest.spyOn(path, "join").mockReturnValue(joinedPath);
            const file = Config.search(configFile, { stop: home, gbl: path.resolve(configDir, "..") });
            expect(file).not.toBe(notExpectedPath);
            expect(file).toBe(expectedPath);
        });
        it("should fail to find a file", async () => {
            const joinedPath = path.join(configDir, configFile);
            jest.spyOn(fs, "existsSync").mockReturnValue(false);
            jest.spyOn(path, "join").mockReturnValue(joinedPath);
            const file = Config.search(configFile, { stop: home });
            expect(file).toBe(null);
        });
        it("should search for and find a file without opts", async () => {
            const joinedPath = path.join(configDir, configFile);
            const expectedPath = path.join(path.resolve(configDir), configFile);
            jest.spyOn(fs, "existsSync").mockReturnValue(true);
            jest.spyOn(path, "join").mockReturnValue(joinedPath);
            const file = Config.search(configFile);
            expect(file).toBe(expectedPath);
        });
        it("should search for and fail to find a file without opts", async () => {
            const joinedPath = path.join(configDir, configFile);
            jest.spyOn(fs, "existsSync").mockReturnValue(false);
            jest.spyOn(path, "join").mockReturnValue(joinedPath);
            const file = Config.search(configFile, { stop: home });
            expect(file).toBe(null);
        });
        it("should throw an error if the previously searched directory is the same is the current", async () => {
            const joinedPath = path.join(configDir, configFile);
            let error;
            jest.spyOn(fs, "existsSync").mockReturnValue(false);
            jest.spyOn(path, "join").mockReturnValueOnce(joinedPath).mockReturnValue(fakeConfigDir);
            jest.spyOn(path, "resolve").mockReturnValue(joinedPath);
            jest.spyOn(path, "parse").mockReturnValueOnce(parsedPath);
            try {
                Config.search(configFile, { stop: home });
            } catch (err) {
                error = err;
            }
            expect(error).toBeDefined();
            expect(error.message).toContain("internal search error");
        });
    });
});
