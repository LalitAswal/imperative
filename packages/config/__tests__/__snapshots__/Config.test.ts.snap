// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config tests api layers get should get the active layer 1`] = `
Object {
  "exists": true,
  "global": false,
  "path": "/home/ah891585/GitHub/imperative/packages/config/__tests__/__resources__/project.config.user.json",
  "properties": Object {
    "defaults": Object {
      "fruit": "fruit.apple",
    },
    "plugins": Array [
      "@zowe/fruit-for-imperative",
    ],
    "profiles": Object {
      "fruit": Object {
        "profiles": Object {
          "apple": Object {
            "properties": Object {
              "color": "red",
            },
            "type": "fruit",
          },
          "orange": Object {
            "properties": Object {
              "color": "orange",
            },
            "type": "fruit",
          },
        },
        "properties": Object {
          "origin": "California",
        },
      },
    },
    "secure": Array [],
  },
  "user": true,
}
`;

exports[`Config tests api layers set should set the current layer 1`] = `
Object {
  "$schema": "fake",
  "defaults": Object {},
  "plugins": Array [],
  "profiles": Object {
    "vegetable": Object {
      "properties": Object {
        "color": "brown",
        "origin": "California",
      },
    },
  },
  "secure": Array [],
}
`;

exports[`Config tests api plugins get should get the plugins 1`] = `
Array [
  "@zowe/fruit-for-imperative",
]
`;

exports[`Config tests api profiles defaultGet should get the default profile 1`] = `
Object {
  "color": "red",
  "origin": "California",
}
`;

exports[`Config tests api profiles get should get a first level profile 1`] = `
Object {
  "origin": "California",
}
`;

exports[`Config tests api profiles get should get a second level profile 1`] = `
Object {
  "color": "red",
  "origin": "California",
}
`;

exports[`Config tests api profiles set should set a first level profile 1`] = `
Object {
  "color": "green",
}
`;

exports[`Config tests api profiles set should set a second level profile 1`] = `
Object {
  "origin": "California",
}
`;

exports[`Config tests api profiles set should set a second level profile 2`] = `
Object {
  "color": "green",
  "origin": "California",
}
`;

exports[`Config tests api profiles set should set a second level profile to a first level profile that doesn't exist 1`] = `
Object {
  "color": "brown",
}
`;

exports[`Config tests api profiles set should set a second level profile to a first level profile that doesn't exist 2`] = `
Object {
  "color": "brown",
}
`;

exports[`Config tests load should load global config 1`] = `
Object {
  "defaults": Object {
    "fruit": "fruit.banana",
    "vegetable": "vegetable.artichoke",
  },
  "plugins": Array [],
  "profiles": Object {
    "fruit": Object {
      "profiles": Object {
        "apple": Object {
          "properties": Object {
            "color": "green",
          },
          "type": "fruit",
        },
        "banana": Object {
          "properties": Object {
            "color": "yellow",
          },
          "type": "fruit",
        },
      },
      "properties": Object {
        "origin": "California",
      },
    },
    "vegetable": Object {
      "profiles": Object {
        "artichoke": Object {
          "properties": Object {
            "color": "green",
          },
          "type": "vegetable",
        },
      },
      "properties": Object {
        "origin": "China",
      },
    },
  },
  "secure": Array [],
}
`;

exports[`Config tests load should load project config 1`] = `
Object {
  "defaults": Object {
    "fruit": "fruit.orange",
  },
  "plugins": Array [],
  "profiles": Object {
    "fruit": Object {
      "profiles": Object {
        "apple": Object {
          "properties": Object {
            "color": "green",
          },
          "type": "fruit",
        },
        "orange": Object {
          "properties": Object {
            "color": "orange",
          },
          "type": "fruit",
        },
      },
      "properties": Object {
        "origin": "California",
      },
    },
  },
  "secure": Array [],
}
`;

exports[`Config tests load should load project user config 1`] = `
Object {
  "defaults": Object {
    "fruit": "fruit.apple",
  },
  "plugins": Array [
    "@zowe/fruit-for-imperative",
  ],
  "profiles": Object {
    "fruit": Object {
      "profiles": Object {
        "apple": Object {
          "properties": Object {
            "color": "red",
          },
          "type": "fruit",
        },
        "orange": Object {
          "properties": Object {
            "color": "orange",
          },
          "type": "fruit",
        },
      },
      "properties": Object {
        "origin": "California",
      },
    },
  },
  "secure": Array [],
}
`;

exports[`Config tests load should load user config 1`] = `
Object {
  "defaults": Object {
    "fruit": "fruit.apple",
    "vegetable": "vegetable.artichoke",
  },
  "plugins": Array [],
  "profiles": Object {
    "fruit": Object {
      "profiles": Object {
        "apple": Object {
          "properties": Object {
            "color": "red",
          },
          "type": "fruit",
        },
        "banana": Object {
          "properties": Object {
            "color": "yellow",
          },
          "type": "fruit",
        },
      },
      "properties": Object {
        "origin": "California",
      },
    },
    "vegetable": Object {
      "profiles": Object {
        "artichoke": Object {
          "properties": Object {
            "color": "green",
          },
          "type": "vegetable",
        },
      },
      "properties": Object {
        "origin": "Mexico",
      },
    },
  },
  "secure": Array [],
}
`;

exports[`Config tests load should merge multiple config files 1`] = `
Object {
  "defaults": Object {
    "fruit": "fruit.apple",
    "vegetable": "vegetable.artichoke",
  },
  "plugins": Array [
    "@zowe/fruit-for-imperative",
  ],
  "profiles": Object {
    "fruit": Object {
      "profiles": Object {
        "apple": Object {
          "properties": Object {
            "color": "red",
          },
          "type": "fruit",
        },
        "orange": Object {
          "properties": Object {
            "color": "orange",
          },
          "type": "fruit",
        },
      },
      "properties": Object {
        "origin": "California",
      },
    },
    "vegetable": Object {
      "profiles": Object {
        "artichoke": Object {
          "properties": Object {
            "color": "green",
          },
          "type": "vegetable",
        },
      },
      "properties": Object {
        "origin": "Mexico",
      },
    },
  },
  "secure": Array [],
}
`;
